umap/distances.py

```diff
@@ -182,6 +182,43 @@ def minkowski_grad(x, y, p=2):
     return result ** (1.0 / p), grad
 
 
+@numba.njit()
+def poincare(u, v):
+    """Poincare distance.
+
+    ..math::
+        \delta (u, v) = 2 \frac{ \lVert  u - v \rVert ^2 }{ ( 1 - \lVert  u \rVert ^2 ) ( 1 - \lVert  v \rVert ^2 ) }
+        D(x, y) = \operatorname{arcosh} (1+\delta (u,v))
+    """
+    sq_u_norm = np.sum(u * u)
+    sq_v_norm = np.sum(v * v)
+    sq_dist = np.sum(np.power(u - v, 2))
```

Comment:
Depending on how fast you want this to run it may be beneficial to compute u-v and then write the for loop to sum up the squares. np.power under numba is definitely slower that looping and using **2.

Reply:
gotcha, yeah i'll do that then